monitor/monitor.go

```diff
@@ -44,54 +45,53 @@ type Monitor struct {
 
 // Run starts monitoring.
 func (m *Monitor) Run(npages int) {
-	log.Info("Starting monitoring traffic")
-	c := bpf.DefaultPerfEventConfig()
-	c.NumPages = npages
-
-	me, err := bpf.NewPerCpuEvents(c)
-	if err != nil {
-		log.WithError(err).Error("Error while starting monitor")
-		return
-	}
-	monitorEvents = me
-
+	eventConfig.NumPages = npages
 	// Dump stat
 	t := time.NewTicker(5 * time.Second)
-	go func(t *time.Ticker) {
-		for {
-			select {
-			case <-t.C:
-				if monitorEvents != nil {
-					m.dumpStat()
-				}
-			}
-		}
-	}(t)
-
-	// Main event loop
 	for {
-		todo, err := monitorEvents.Poll(pollTimeout)
-		if err != nil {
-			log.WithError(err).Error("Error in Poll")
-			if err == syscall.EBADF {
-				break
+		select {
+		case <-t.C:
+			if monitorEvents != nil {
+				m.dumpStat()
 			}
 		}
-		if todo > 0 {
-			if err := monitorEvents.ReadAll(m.receiveEvent, m.lostEvent); err != nil {
-				log.WithError(err).Warn("Error received while reading from perf buffer")
-			}
+	}
+}
+
+func purgeInactiveListeners() {
+	meta := &payload.Meta{Size: uint32(0)}
+	metaBuf, err := meta.MarshalBinary()
+	if err != nil {
+		log.WithError(err).Fatal("meta encode")
+	}
+
+	var next *list.Element
+	for e := listeners.Front(); e != nil; e = next {
+		client := e.Value.(net.Conn)
+		next = e.Next()
+		if _, err := client.Write(metaBuf); err != nil {
+			log.WithError(err).Warn("metadata write failed; removing client")
+			client.Close()
+			listeners.Remove(e)
+			continue
 		}
 	}
 }
 
+func numListeners() int {
+	mutex.Lock()
+	defer mutex.Unlock()
+	return listeners.Len()
+}
+
 // dumpStat prints out the monitor status in JSON.
 func (m *Monitor) dumpStat() {
 	c := int64(monitorEvents.Cpus)
 	n := int64(monitorEvents.Npages)
 	p := int64(monitorEvents.Pagesize)
 	l, u := monitorEvents.Stats()
-	ms := models.MonitorStatus{Cpus: c, Npages: n, Pagesize: p, Lost: int64(l), Unknown: int64(u)}
+	num := int64(numListeners())
+	ms := models.MonitorStatus{Cpus: c, Npages: n, Pagesize: p, Lost: int64(l), Unknown: int64(u), Nlisteners: num}
```

Comment:
Expand this to multiple lines, it will become easier to read and you don't need to store `num` first.